File: boj/9012.java
// https://www.acmicpc.net/problem/9012

import java.io.*;

public class Main {
    public static void main(String[] args) throws Exception {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));

        int n = Integer.parseInt(br.readLine());
        for (int i=0; i<n; i++) {
            if (isVPS(br.readLine())) {
                bw.write("YES\n");
            } else {
                bw.write("NO\n");
            }
        }
        bw.flush();
    }

    static boolean isVPS(String s) {
        int count = 0;
        for (char c : s.toCharArray()) {
            if (c == '(') {
                count++;
            }
            else {
                if (count == 0) return false;
                count--;
            }
        }

        return count == 0;
    }
}

File: boj/kotlin/9012.kt
// https://www.acmicpc.net/problem/9012

import java.io.*;

fun main(args: Array<String>) = with(BufferedReader(InputStreamReader(System.`in`))) {
    val bw = BufferedWriter(OutputStreamWriter(System.out))
    repeat(readLine().toInt()) {
        if (isVPS(readLine())) bw.write("YES\n") else bw.write("NO\n")
    }
    bw.flush()
}

fun isVPS(s: String) : Boolean {
    var count = 0
    s.toCharArray().forEach {
        when(it) {
            '(' -> count++
            ')' -> if (count == 0) return false else count--
        }
    }
    return count == 0
}
